servicemanager/servicemanager_test.go

```diff
@@ -0,0 +1,605 @@
+
+	server := &http.Server{Addr: ":9000", Handler: http.FileServer(http.Dir(fileServerDir))}
+
+	go func() {
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			log.Errorf("Can't serve http server: %s", err)
+		}
+	}()
+
+	time.Sleep(1 * time.Second)
+
+	defer server.Close()
+
+	if err = sm.InstallService(servicemanager.ServiceInfo{ServiceID: serviceID, AosVersion: 1},
+		"http://:9000/downloadImage", fileInfo); err != nil {
+		t.Errorf("Can't install service from remote domain: %s", err)
+	}
+
+	os.RemoveAll(path.Join(fileServerDir, "downloadImage"))
+
+	// test version missmatch
+	serviceID = "testID1"
+
+	if serviceURL, fileInfo, err = prepareService("SomeContnet"); err != nil {
+		t.Fatalf("Can't prepare test service: %s", err)
+	}
+
+	serviceStorage.Services = append(serviceStorage.Services, servicemanager.ServiceInfo{
+		AosVersion: 1,
+		ServiceID:  serviceID,
+	})
+
+	if err = sm.InstallService(servicemanager.ServiceInfo{ServiceID: serviceID, AosVersion: 1},
+		serviceURL, fileInfo); err == nil {
+		t.Error("Should be error version mismatch")
+	} else if !errors.Is(err, servicemanager.ErrVersionMismatch) {
+		t.Errorf("Should be error version mismatch, but have: %s", err)
+	}
+
+	os.RemoveAll(serviceURL)
+
+	// check incorrect check sum
+	serviceID = "testID2"
+
+	if serviceURL, fileInfo, err = prepareService("SomeContnet"); err != nil {
+		t.Fatalf("Can't prepare test service: %s", err)
+	}
+
+	fileInfo.Sha256 = []byte{0}
+	if err = sm.InstallService(servicemanager.ServiceInfo{ServiceID: serviceID, AosVersion: 1},
+		serviceURL, fileInfo); err == nil {
+		t.Error("Should be error")
+	}
+
+	// untar error
+	notTarFile := path.Join(tmpDir, "notTar")
+	if err := ioutil.WriteFile(notTarFile, []byte("testContent"), 0o600); err != nil {
+		t.Errorf("Can't create file: %s", err)
+	}
+
+	if fileInfo, err = image.CreateFileInfo(context.Background(), notTarFile); err != nil {
+		t.Errorf("Can't create file info: %s", err)
+	}
+
+	if err = sm.InstallService(servicemanager.ServiceInfo{ServiceID: serviceID, AosVersion: 1},
+		"file://"+notTarFile, fileInfo); err == nil {
+		t.Error("Should be error can't untar")
+	}
+
+	// check download error
+	if err = sm.InstallService(servicemanager.ServiceInfo{ServiceID: serviceID, AosVersion: 1},
+		"url://notexist", fileInfo); err == nil {
+		t.Error("Should be error")
+	}
+
+	// test incorrect url
+	if err = sm.InstallService(servicemanager.ServiceInfo{ServiceID: serviceID, AosVersion: 1},
+		"\n", fileInfo); err == nil {
+		t.Error("Should be error")
+	}
+
+	os.RemoveAll(serviceURL)
+
+	// test get service error
+	if serviceURL, fileInfo, err = prepareService("SomeContnet"); err != nil {
+		t.Fatalf("Can't prepare test service: %s", err)
+	}
+
+	if err = sm.InstallService(servicemanager.ServiceInfo{ServiceID: errorGetServicID, AosVersion: 1},
+		serviceURL, fileInfo); err == nil {
+		t.Error("Should be can't add service")
```

Comment:
Service should not be installed
Should be error
Should be error: can't install service

Reply:
Decided to use Should be error: can't install service